### ISSUE TITLE
Benchmarking code still uses library()

### ISSUE THREAD

**Issue by bcow** [PecanProject/pecan#1592]
Need to make sure that all benchmarking code no longer loads full packages at the beginning of functions as part of general cleanup.

**Comment by infotroph** [comment]
Unstaling: Still on the list and would be a good issue for a first-time contributor.

**Comment by moki1202** [comment]
@infotroph can you please explain the issue in detail to me ? I would really like to work on this issue.

**Comment by ashiklom** [comment]
The gist is that using `library(<package>)` to load a package inside of another package's function definition is bad practice. The recommended way to use functions from another package is `package::function`, or, in some cases, via a Roxygen directive like `#' @importFrom package function`. See https://r-pkgs.org/namespace.html#imports for more details on R Namespaces.

**Comment by moki1202** [comment]
I manually checked all the R files and couldn't find this issue. can someone ping the file/files to be fixed? it would be greatly appreciated.

**Comment by infotroph** [comment]
## First a terminology sidebar

I'll use "plain call" here to mean calling a function just by its name (`filter()`), as opposed to a "namespaced call" where we call it by the name of the function *and* the name of the package the function lives in (`dplyr::filter()`). See the link @ashiklom posted above for a lot more detail on exactly what the difference is and why it matters.

## Which files need editing?

The issue description refers to the benchmarking module, so we should be looking in the `modules/benchmark/` directory.
Since `library` is only a problem inside package functions  (it's fine in, say, a vignette showing usage examples), we know we can focus only on the function definitions, which all live in the `R/` subdirectory.

Knowing this, we can search the files in `modules/benchmark/R` for calls to the `library()` function. I like working in a Unix shell, so I'll use `grep`, but you can do the same with the find-in-multiple-file feature of VScode/Sublime/Atom/your-favorite-editor-here:

```{sh}
$> grep -R 'library' modules/benchmark/R
modules/benchmark/R/pecan_bench.R:  ## MCD2: in R you use the function taylor.diagram in the 'plotrix' library to make the plot. you
modules/benchmark/R/load_data.R:  library(PEcAn.utils)
modules/benchmark/R/load_data.R:  library(PEcAn.benchmark)
modules/benchmark/R/load_data.R:  library(lubridate)
modules/benchmark/R/load_data.R:  library(udunits2)
modules/benchmark/R/load_data.R:  library(dplyr)
```

Looks like load_data.R is the only file that needs fixing! But it has five different `library()` calls to be removed, so there's plenty to do.

(One of those five can be removed immediately with no further work! Can you find which one? Hint: A function never needs to load its own package's namespace).


## Which function calls need namespaces added?

For every `library(some_pkg)` call to be removed, the procedure is the same: Find all the places `load_data` uses a function from `some_pkg`, convert any of those places that are plain calls into namespaced calls by adding `some_pkg::` right before the function name. 

There are many possible ways to do the "find all the places..."" step, ranging from tedious manual search up through very fancy automated diagnostics. For today, probably the easiest is to refer to `modules/benchmark/tests/Rcheck_reference.log`, which is a saved copy of all the ways R's compile-time package checks complain about the PEcAn.benchmark package -- every line of this file is a problem we're currently ignoring but will need to fix eventually. By searching that file for the phrase "no visible global function definition" and focusing only on messages that come from the `load_data` function, I get a list of seven plain calls that need attention:

```
load_data: no visible global function definition for ‘get_key’
load_data: no visible global function definition for ‘get_token’
load_data: no visible global function definition for ‘convert_file’
load_data: no visible global function definition for
	‘misc.are.convertible’
load_data: no visible global function definition for ‘misc.convert’
load_data: no visible global function definition for ‘one_of’
load_data: no visible global function definition for ‘%>%’
```

Six of of these should be edited to add namespace prefixes and it's "only" a matter of finding out which package each one comes from. Note that several of these plain calls are functions from packages OTHER THAN the five we set out to remove, which means they still ought to be fixed but that removing `library` calls won't make them break any worse than they are.

The seventh one is the dplyr/magrittr pipe `%>%`, which can't be called with a namespace (because ```mtcars magrittr::`%>%` head()``` would be silly!) so it needs to be imported into the package namespace with an `@importFrom`, like we've [previously done for the PEcAn.DB package](https://github.com/PecanProject/pecan/blob/develop/base/db/R/PEcAn.DB-package.R#L7).


BTW, You can also use `lintr` to check for function calls that are missing namespaces, e.g. something like: 

```{r}
lintr::lint("modules/benchmark/R/load_data.R", linters = lintr::object_usage_linter)
```

This should give you the same list of plain calls as above, but is easy to re-check after edits to make sure you really fixed things, and unlike `Rcheck_reference.log` it also reports line numbers.


## Which package's namespace does this function come from?

There are many possible ways to do this as well; in many cases you might know already just from using the package in other contexts. In cases were you're less sure, some strategies include:

If you have all the possibly relevant packages installed in R on your machine, searching help with `??` is often all you need.

```
??one_of
# dplyr::reexports        Objects exported from other packages
#   Aliases: one_of
# rex::character_class    Create character classes
#  Aliases: one_of, none_of
# tidyr::reexports        Objects exported from other packages
#  Aliases: one_of
# tidyselect::one_of      Superseded selection helpers
#  Aliases: one_of
```

==> When `load_data` was written, `one_of` was part of dplyr, but apparently that has since changed! It should now be written `tidyselect::one_of`, and possibly the code should be refactored to use `tidyselect::any_of` or `tidyselect::all_of` instead. 

If you suspect a function is defined elsewhere in PEcAn, searching the exports lists of every package can help:

```{sh}
cd /path/to/my/copy/of/pecan
grep --include='NAMESPACE' -R 'misc.convert' .
# 
# ./base/utils/NAMESPACE:export(misc.convert)
```

==> `misc.convert` lives in the PEcAn.utils package, meaning that every `misc.convert(...)` in the `load_data` code should be converted to `PEcAn.utils::misc.convert(...)`.

If all else fails, sometimes a web search for the function name is enough to identify the package it comes from.

**Comment by moki1202** [comment]
@infotroph I can't thank you enough for your help!! I am pretty sure we can remove the library(PEcAn.benchmark ) immediately. I also found that functions "get_key", "get_token" and "convert-file" are part of BrownDog package. please correct me if I am wrong.

**Comment by infotroph** [comment]
yep, correct on all counts! Turns out the Brown Dog package is loaded with `require` inside an if statement (line 47) -- this should be converted to `requireNamespace(bd, quietly = TRUE)` and `bd` should be added to the `Suggests` section of modules/benchmark/DESCRIPTION

**Comment by infotroph** [comment]
Fixed by #2771. Thanks, @moki1202!
